### ISSUE TITLE
How can I step debug prometheus

### ISSUE THREAD

**Issue by maklemenz** [prometheus/prometheus#2298]
**What did you do?**

I tried the intellij golang debugger.
I tried to compile prometheus using go build.

**What did you expect to see?**

I expected to be able to step debug through the code

**What did you see instead? Under which circumstances?**

```
# command-line-arguments
cmd/prometheus/main.go:67: undefined: parse
cmd/prometheus/main.go:72: undefined: cfg in cfg.printVersion
cmd/prometheus/main.go:86: undefined: cfg in cfg.localStorageEngine
cmd/prometheus/main.go:88: undefined: cfg in cfg.storage
cmd/prometheus/main.go:93: undefined: cfg in cfg.localStorageEngine
cmd/prometheus/main.go:97: undefined: cfg in cfg.remote
cmd/prometheus/main.go:112: undefined: cfg in cfg.notifier
cmd/prometheus/main.go:114: undefined: cfg in cfg.queryEngine
cmd/prometheus/main.go:123: undefined: cfg in cfg.web
cmd/prometheus/main.go:126: undefined: cfg
cmd/prometheus/main.go:126: too many errors

Process finished with exit code 2
```

**Environment**

Ubuntu 16.04 LTS

**Comment by juliusv** [comment]
How did you invoke `go build`? The error reads like you only invoked it on one file (`main.go`), not the entire directory / package (`cmd/prometheus`), since it cannot find e.g. `parse()`, which is defined in the same directory in `config.go`.

**Comment by maklemenz** [comment]
@juliusv yes I called it on the main.go

I guess my whole project setup is messed up:

```
mk@mk:~/git/prometheus$ go build cmd/prometheus/main.go 
# command-line-arguments
cmd/prometheus/main.go:67: undefined: parse
cmd/prometheus/main.go:72: undefined: cfg in cfg.printVersion
cmd/prometheus/main.go:86: undefined: cfg in cfg.localStorageEngine
cmd/prometheus/main.go:88: undefined: cfg in cfg.storage
cmd/prometheus/main.go:93: undefined: cfg in cfg.localStorageEngine
cmd/prometheus/main.go:97: undefined: cfg in cfg.remote
cmd/prometheus/main.go:112: undefined: cfg in cfg.notifier
cmd/prometheus/main.go:114: undefined: cfg in cfg.queryEngine
cmd/prometheus/main.go:123: undefined: cfg in cfg.web
cmd/prometheus/main.go:126: undefined: cfg
cmd/prometheus/main.go:126: too many errors

mk@mk:~/git/prometheus$ go build cmd/prometheus/
can't load package: package cmd/prometheus: open /usr/lib/go-1.6/src/cmd/prometheus: no such file or directory

mk@mk:~/git/prometheus$ go build cmd/prometheus/*
# command-line-arguments
cmd/prometheus/main.go:204: cannot use notifier (type *notifier.Notifier) as type "github.com/prometheus/client_golang/prometheus".Collector in argument to "github.com/prometheus/client_golang/prometheus".MustRegister:
	*notifier.Notifier does not implement "github.com/prometheus/client_golang/prometheus".Collector (wrong type for Collect method)
		have Collect(chan<- "github.com/prometheus/prometheus/vendor/github.com/prometheus/client_golang/prometheus".Metric)
		want Collect(chan<- "github.com/prometheus/client_golang/prometheus".Metric)
cmd/prometheus/main.go:282: cannot use []"github.com/prometheus/common/model".LabelSet literal (type []"github.com/prometheus/common/model".LabelSet) as type []"github.com/prometheus/prometheus/vendor/github.com/prometheus/common/model".LabelSet in field value

```

**Comment by juliusv** [comment]
Try `go build ./cmd/prometheus`. `go build` normally expects the name of a package to build (in this case `./cmd/prometheus` or canonically `github.com/prometheus/prometheus/cmd/prometheus`), although it also allows you to specify individual files for one-off use cases.

**Comment by maklemenz** [comment]
well, it would have been too easy if it would just work
```
mk@mk:~/git/prometheus$ go build ./cmd/prometheus/
# _/home/mk/git/prometheus/cmd/prometheus
cmd/prometheus/main.go:204: cannot use notifier (type *notifier.Notifier) as type "github.com/prometheus/client_golang/prometheus".Collector in argument to "github.com/prometheus/client_golang/prometheus".MustRegister:
	*notifier.Notifier does not implement "github.com/prometheus/client_golang/prometheus".Collector (wrong type for Collect method)
		have Collect(chan<- "github.com/prometheus/prometheus/vendor/github.com/prometheus/client_golang/prometheus".Metric)
		want Collect(chan<- "github.com/prometheus/client_golang/prometheus".Metric)
cmd/prometheus/main.go:282: cannot use []"github.com/prometheus/common/model".LabelSet literal (type []"github.com/prometheus/common/model".LabelSet) as type []"github.com/prometheus/prometheus/vendor/github.com/prometheus/common/model".LabelSet in field value
```

Running it is a whole other story:
```
mk@mk:~/git/prometheus$ go run ./cmd/prometheus
go run: no go files listed
```

**Comment by juliusv** [comment]
Hmm, `go build ./cmd/prometheus` works for me.

The error you're getting reads like it's using a wrong version of the Prometheus client library that doesn't fit together with the Prometheus code base, although it should only be using the correct one that is vendored in Prometheus itself (under `./vendor/...`). Not sure what is going on there in your setup... what Go version are you using? Maybe one that doesn't know about vendoring by default yet?

**Comment by maklemenz** [comment]
I found a way.

Make sure prometheus is in `$GOPATH/src/github.com/prometheus/prometheus`.
If you use intellij (not goland eap) use a `Go Application (kind file)` and not `Go Single File` as run configuration
Make sure your go compiler is up to date

While `go build ./cmd/prometheus` will work, `go run ./cmd/prometheus` will not. `go run  ./cmd/prometheus/main.go` also does not work as expected. `go run` is the way that intellij will use to start and debug the application so that just has to work.
As a workaround (definitely not a solution) I had to copy the contents from `cmd/prometheus/config.go` into `cmd/prometheus/main.go` and remove `cmd/prometheus/config*`
Then go run `$GOPAH/src/github.com/prometheus/prometheus/cmd/prometheus/main.go` finally worked.
